services/contrib/FactsServiceAgent/facts_service/agent.py

```diff
@@ -0,0 +1,433 @@
+    :rtype: `FactsService`
+    """
+
+    if isinstance(config_path, dict):
+        config_dict = config_path
+    else:
+        config_dict = utils.load_config(config_path)
+
+    # Gather all settings from configuration into kwargs.
+    # This ensures that settings common to all historians
+    # are passed to BaseHistorian.
+    utils.update_kwargs_with_config(kwargs, config_dict)
+    return FactsService(**kwargs)
+
+
+class FactsService(BaseHistorian):
+    """
+    Sends trend data to Ecorithm's Facts Service using HTTP PUT requests.
+    """
+
+    def __init__(self, facts_service_parameters={}, building_parameters={},
+                 **kwargs):
+        # The publish_to_historian function will run in a
+        # separate thread unless you change this to True.
+        # Unless you need to interact with the VOLTTRON platform
+        # leave this unchanged.
+        kwargs["process_loop_in_greenlet"] = False
+        super(FactsService, self).__init__(**kwargs)
+
+        self._facts_service_base_api_url = None
+        self._facts_service_username = None
+        self._facts_service_password = None
+        self._building_id = None
+        self._topic_building_mapping = None
+        self._db_connection = None
+        self._db_is_alive = False
+
+        # The base historian handles the interaction with the
+        # configuration store.
+        config = {
+            "facts_service_parameters": facts_service_parameters,
+            "building_parameters": building_parameters
+        }
+        # Add our settings to the base historians default settings.
+        self.update_default_config(config)
+
+    def configure(self, configuration):
+        # The base historian will call this whenever the
+        # Historian is reconfigured in the main process thread.
+
+        # If the Historian is already running historian_teardown
+        # will be called before this is called and
+        # historian_setup will be called afterwards.
+
+        facts_service_parameters = configuration["facts_service_parameters"]
+        building_parameters = configuration["building_parameters"]
+
+        if not isinstance(facts_service_parameters, dict):
+            _log.warning(
+                "Supplied facts_service_parameters is not a dict, ignored"
+            )
+            facts_service_parameters = {}
+
+        if not isinstance(building_parameters, dict):
+            _log.warning("Supplied building_parameters is not a dict, ignored")
+            building_parameters = {}
+
+        self._facts_service_base_api_url = \
+            facts_service_parameters.get("base_api_url")
+        self._facts_service_username = facts_service_parameters.get("username")
+        self._facts_service_password = facts_service_parameters.get("password")
+        self._building_id = building_parameters.get("building_id")
+        self._topic_building_mapping = \
+            building_parameters.get("topic_building_mapping")
```

Comment:
If both _building_id and _topic_building_mapping are None (they are both missing) this will cause a problem in publish to historian on line 182. _topic_building_mapping should default to {}. You can do that with a second argument to get.